GafferScene : Add ImageToPoints node

johnhaddon opened the pull request:
I made one of these for Nuke many many years ago, and it was a fun little node to have. Someone at Image Engine recently mentioned it'd be nice to have something for Gaffer too, so I spent a bit of friday/evening/train time putting this together.

danieldresser-ie commented on src/GafferScene/ImageToPoints.cpp:
```diff
@@ -0,0 +1,522 @@
+
+	// Strip out points below the alpha threshold in a final serial step.
+	// We can't easily do this in the step above because we don't know the
+	// point count or order until after we've visited all tiles.
```
Create large dense buffers for all channels being converted before we figure out the data layout seems pretty wasteful of performance - I strongly suspect we could do a lot better with a prepass that visits all tiles of alpha only, and then a second pass that puts all data directly into the right place.  I don't see anything that would prevent us from doing that later if we decide that performance is important for this node though.

johnhaddon replied:
That approach would have very unpredictable performance when the cache is insufficient to hold all tiles for all images in the graph. Many nodes (ImageReader and all ColorProcessors to name the obvious ones) compute several channels at once, and this approach risks doing all that work twice if the top-left tiles are evicted before we begin the second pass.

The other consideration is that the serial step is actually pretty minimal compared to the typical cost of generating the input image itself. Here are some stats for the conversion of an EXR input image coming off a local SSD, straight out of an ImageReader and into an ImageToPoints node :

- ImageReader : 0.108s
- ImageToPoints : 0.123s (including ImageReader time)
- ImageToPoints with `ignoreTransparent` on : 0.133 (including ImageReader time)

So we're looking at about 8% overhead, and as image generation time increases, that overhead will only diminish. I think that's a pretty good trade-off against unpredictable performance from cache misses in more complex cases.

But there is an obvious optimisation that I missed - if `ignoreTransparent` is on, we don't need to allocate a buffer for the whole display window, because we know everything outside the data window will be discarded later. I implemented that in https://github.com/GafferHQ/gaffer/pull/5177/commits/2eed153bbe0ecfefc951e425e89a2d4201b2aa5d, for a whopping 93% reduction in runtime for a best-case scenario of a large image with a small area of text.

There are still potential optimisations for the serial step - we could do a prepass to build a vector of indices of non-zero-alpha pixels, limiting the memory accesses when doing the actual channel processing. And we could parallelize over the primitive variables. But I don't really see any need to explore that at this point.

danieldresser-ie reviewed (commented):
My only concern about the behaviour this is establishing is the use of displayWindow.  It seems undesirable in most cases.  I don't see a lot of reason why you wouldn't want to include overscan outside the display window.  Even worse is the case where you're using ignoreTransparent with a small crop - as far as I can tell, you would allocate large buffers for the whole image, only to spend time later scanning through all the zeros and discarding them?  Is there any reason not to just use the data window?

johnhaddon commented:
> My only concern about the behaviour this is establishing is the use of displayWindow. It seems undesirable in most cases. I don't see a lot of reason why you wouldn't want to include overscan outside the display window. 

[Anatomy of an Image](http://www.gafferhq.org/documentation/1.2.0.2/WorkingWithImages/AnatomyOfAnImage/index.html), which was derived from the EXR spec, is pretty clear that the display window is the boundary of the visible image. So I feel pretty strongly that that is what ImageToPoints should convert. Using the data window to include overscan pixels introduces ambiguity about the interpretation of pixels a) inside the display window and outside the data window b) outside the display window _and_ outside the dataWindow, but within the union of the two. I see no need to guess the user's intentions for those when there's already an unambiguous definition of the image area, and a Crop node trivially allows the user to change it.

> Even worse is the case where you're using ignoreTransparent with a small crop - as far as I can tell, you would allocate large buffers for the whole image, only to spend time later scanning through all the zeros and discarding them? 

Yep, that's a good point. Implemented a trivial optimisation for that in https://github.com/GafferHQ/gaffer/pull/5177/commits/2eed153bbe0ecfefc951e425e89a2d4201b2aa5d.

I should point out that I've rebased this to fix a conflict in Changes.md and to use `#pragma once`, but otherwise the original commits are unchanged. Only the last commit with the optimisation is genuinely new.

johnhaddon commented:
> and a Crop node trivially allows the user to change it.

I will concede here that it's harder than it should be to extend the display window to meet the data window at present, but that's easily addressed with an additional `Crop::AreaSource` for the union of the two.

johnhaddon commented:
Pushed one more small commit, to document the `displayWindow` semantics and the availability of the Crop node : https://github.com/GafferHQ/gaffer/pull/5177/commits/b71c61639c4d8b062f6bc5d220ad6bc6c320362b.

danieldresser-ie commented:
Approved.  I still think a toggle for people who think in data windows would be nice, but it could be added later.

murraystevenson commented on python/GafferSceneUI/ImageToPointsUI.py:
```diff
@@ -0,0 +1,170 @@
+
+			"description",
+			"""
+			Threshold used to exclude pixels from the pointcloude when `ignoreTransparent`
```
Ye olde pointcloude shoppe.

johnhaddon replied:
Also fixed in https://github.com/GafferHQ/gaffer/pull/5177/commits/a771ceb6acd146ad66ab1336f7969510a03a75c3.

Maybe this is how we should encourage long-running shows to be a bit more ambitious in their upgrade schedules - regress the language in the documentation gradually back towards the 16th century with each point release :)

murraystevenson reviewed (commented):
Nice one! It's a fun node to play with. I've made inline suggestions about the use of "pointcloud" in the node and plug descriptions, and a potential improvement to the widget used on `widthChannel`, otherwise it looks good to include in the upcoming `1.2.1.0` release. :+1:

murraystevenson commented on src/GafferScene/ImageToPoints.cpp:
```diff
@@ -0,0 +1,530 @@
+
+	if( !widthChannel.empty() and !haveWidthMapping )
+	{
+		throw IECore::Exception( "Width channels is missing from the input image" );
```
"Width channel is missing from the input image"

johnhaddon replied:
Also fixed in https://github.com/GafferHQ/gaffer/commit/a771ceb6acd146ad66ab1336f7969510a03a75c3.

johnhaddon commented on python/GafferSceneUI/ImageToPointsUI.py:
```diff
@@ -0,0 +1,170 @@
+			The channel used to provide per-point width values for the pointcloud.
+			""",
+
+			"plugValueWidget:type", "GafferImageUI.ChannelPlugValueWidget",
```
Good point. The ChannelPlugValueWidget already has a different mechanism for adding menu items, so I used that to add `None` in https://github.com/GafferHQ/gaffer/pull/5177/commits/a771ceb6acd146ad66ab1336f7969510a03a75c3. Longer term I'd like to look at cleaning up and unifying the metadata for these similar widgets, but I don't think adding new metadata to ChannelPlugValueWidget now will necessarily aid that (it already has two somewhat overlapping mechanisms).

johnhaddon commented on python/GafferSceneUI/ImageToPointsUI.py:
```diff
@@ -0,0 +1,170 @@
+
+	"description",
+	"""
+	Converts an image into a 3d pointcloud, with a point for each pixel
```
Also done in https://github.com/GafferHQ/gaffer/pull/5177/commits/a771ceb6acd146ad66ab1336f7969510a03a75c3
